lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -428,221 +397,289 @@ def create_image(module, ec2):
 
-    # Get all associated snapshot ids before deregistering image otherwise this information becomes unavailable
+    # Get all associated snapshot ids before deregistering image otherwise this information becomes unavailable.
     snapshots = []
-    if hasattr(img, 'block_device_mapping'):
-        for key in img.block_device_mapping:
-            snapshots.append(img.block_device_mapping[key].snapshot_id)
+    if 'BlockDeviceMappings' in image:
+        for mapping in image.get('BlockDeviceMappings'):
+            snapshots.append(mapping.get('SnapshotId'))
 
-    # When trying to re-delete already deleted image it doesn't raise an exception
-    # It just returns an object without image attributes
-    if hasattr(img, 'id'):
+    # When trying to re-deregister an already deregistered image it doesn't raise an exception, it just returns an object without image attributes.
+    if 'ImageId' in image:
         try:
-            params = {'image_id': image_id,
-                      'delete_snapshot': delete_snapshot}
-            ec2.deregister_image(**params)
-        except boto.exception.BotoServerError as e:
-            module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
+            res = connection.deregister_image(ImageId=image_id)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Error deregistering image - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
     else:
-        module.exit_json(msg="Image %s has already been deleted" % image_id, changed=False)
+        module.exit_json(msg="Image %s has already been deregistered." % image_id, changed=False)
 
-    # wait here until the image is gone
-    img = ec2.get_image(image_id)
+    image = get_image_by_id(module, connection, image_id)
     wait_timeout = time.time() + wait_timeout
-    while wait and wait_timeout > time.time() and img is not None:
-        img = ec2.get_image(image_id)
+
+    while wait and wait_timeout > time.time() and image is not None:
+        image = get_image_by_id(module, connection, image_id)
         time.sleep(3)
+
     if wait and wait_timeout <= time.time():
-        # waiting took too long
-        module.fail_json(msg="timed out waiting for image to be deregistered/deleted")
+        module.fail_json(msg="Timed out waiting for image to be deregistered.")
+
+    exit_params = {'msg': "AMI deregister operation complete.", 'changed': True}
 
-    # Boto library has hardcoded the deletion of the snapshot for the root volume mounted as '/dev/sda1' only
-    # Make it possible to delete all snapshots which belong to image, including root block device mapped as '/dev/xvda'
     if delete_snapshot:
         try:
             for snapshot_id in snapshots:
-                ec2.delete_snapshot(snapshot_id)
-        except boto.exception.BotoServerError as e:
+                connection.delete_snapshot(snapshot_id)
```

Comment:
I think this needs to be SnapshotId=snapshot_id
```
2017-09-20 14:11:58     "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_xsv9t058/ansible_module_ec2_ami.py\", line 686, in <module>\n    main()\n  File \"/tmp/ansible_xsv9t058/ansible_module_ec2_ami.py\", line 676, in main\n    deregister_image(module, connection)\n  File \"/tmp/ansible_xsv9t058/ansible_module_ec2_ami.py\", line 547, in deregister_image\n    connection.delete_snapshot(snapshot_id)\n  File \"/usr/local/lib/python3.5/dist-packages/botocore/client.py\", line 310, in _api_call\n    \"%s() only accepts keyword arguments.\" % py_operation_name)\nTypeError: delete_snapshot() only accepts keyword arguments.\n",
2017-09-20 14:11:58     "module_stdout": "",
2017-09-20 14:11:58     "msg": "MODULE FAILURE",
2017-09-20 14:11:58     "rc": 0
```

Reply:
fixed